core/headerchain.go

```diff
@@ -337,22 +365,22 @@ func (hc *HeaderChain) GetTd(hash common.Hash) *big.Int {
 
 // WriteTd stores a block's total difficulty into the database, also caching it
 // along the way.
-func (hc *HeaderChain) WriteTd(hash common.Hash, td *big.Int) error {
-	if err := WriteTd(hc.chainDb, hash, td); err != nil {
+func (hc *HeaderChain) WriteTd(hash common.Hash, number uint64, td *big.Int) error {
+	if err := WriteTd(hc.chainDb, hash, number, td); err != nil {
 		return err
 	}
 	hc.tdCache.Add(hash, new(big.Int).Set(td))
 	return nil
 }
 
-// GetHeader retrieves a block header from the database by hash, caching it if
-// found.
-func (hc *HeaderChain) GetHeader(hash common.Hash) *types.Header {
+// getHeader retrieves a block header from the database by hash and number,
+// caching it if found.
+func (hc *HeaderChain) getHeader(hash common.Hash, number uint64) *types.Header {
```

Comment:
Yeah, that's why I also suggested it. Since it requires explicitly adding th eby hash, you got to think over all the use cases and only use hash where the number is truly unavailable :)